src/NServiceBus.Core/PreparedEndpoint.cs

```diff
@@ -0,0 +1,135 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Features;
+    using Installation;
+    using ObjectBuilder;
+    using Settings;
+    using Transport;
+
+    /// <summary>
+    /// A prepared endpoint using an external container.
+    /// </summary>
+    public class PreparedEndpoint
+    {
+        internal PreparedEndpoint()
+        {
+            MessageSession = new UninitializedMessageSession();
+        }
+
+        /// <summary>
+        /// The message session (can only be used once the endpoint have been started).
+        /// </summary>
+        public IMessageSession MessageSession { get; private set; }
```

Comment:
is this used somewhere?

Reply:
Yes this is important to be able to register something early in the container that allows to point to a message session so that you don't have to mutate the container anymore

Reply:
do you have an example where this is used? Shouldn't the `IStartableEndpoint` then also expose a session for consistency? Your scenario also seems potentially risky in case users use the property too early and then suddenly start referencing the dummy session in case they don't fetch the session lazily?

Reply:
@timbussmann here is an example https://github.com/Particular/docs.particular.net/pull/4475/files#diff-aedd8ef13f44709736cc30d622704f26R11

Reply:
The dummy session throws if it is resolved too early

Reply:
>Shouldn't the IStartableEndpoint then also expose a session for consistency? 

For consistency, I guess we could consider it but it would be a bit of a false promise since eg. a MVC user

https://docs.particular.net/samples/web/send-from-mvc-controller/#wire-up-autofac

could register the session before start but `.Create` already does resolves so we would still break immutable containers like SimpleInjector? 

I would call YAGNI on that for now?

Reply:
Maybe it could be made a bit more intention revealing by making it a method? So

```
public Func<IMessageSession> MessageSessionProvider { get; private set; }
```

or something similar. This enforces factory bindings

Reply:
I like this. @SzymonPobiega what do you think?

Reply:
👍 for this

Reply:
Ok, I’ll make the change

Reply:
@danielmarbach was https://github.com/Particular/NServiceBus/pull/5411/commits/5735d9b8f6d406633a1d80ae68b7000b6813fc9d what you had in mind?

Reply:
Yes exactly

Reply:
👍